# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [92]:
# Imports section
import pandas as pd
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import PolynomialFeatures

## 1. Loading the dataset

In [93]:
# Using pandas load the dataset
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [94]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data.drop('Size nm^3',axis=1)
y = data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [96]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = data.loc[[0]].drop(columns=['Size nm^3'])
sample_pred = model.predict(sample)
print("Sample actual size: ", data.iloc[0]['Size nm^3'])
print("Sample predicted size: ", sample_pred[0])

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("Model score: ", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(coefficients, intercept)

Sample actual size:  624474.2571
Sample predicted size:  664984.8963043336
Model score:  0.8552472077276096
[ 866.14641337 1032.69506649] -409391.4795834082


Write the linear equation of a slime: (example equation: $E = mc^2$)

$y=866.15+1032.7x-409391$

Report on score and explain meaning: The $R^2$ score describes how well the linear regression model predicted the labels based off of how far the predicted value was from the actual value (in other words, the residuals). More mathemtically and precisely, it is equal to $(1-(\frac{u}{v}))$, where $u$ is the residual sum of squares and $v$ is the total sum of squares.

## 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kfold = KFold(n_splits = 5, shuffle=True, random_state=42)
scores = cross_val_score(model, X, y, cv=kfold)

# Report on their finding and their significance
print("Scores:", scores)
print("Average score: ", np.mean(scores))
print("Standard deviation: ", np.std(scores))

Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Average score:  0.8597294202684644
Standard deviation:  0.01838773713930639


Write findings here: The mean accuracy of the cross validation accuracy is 0.85973 and the standard deviation is 0.01839. We can see that the cross validation accuracy is very close to the R2 value reported by model.score, and the standard deviation is quite small. This is significant because it shows that our model gives pretty consistent results across different folds, and that these results are decently high.

## 5. Using Polynomial Regression

In [90]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
pr = PolynomialFeatures(degree=2)
X_poly = pr.fit_transform(X)

# Perform k-fold cross validation (as above)
kfold = KFold(n_splits = 5, shuffle=True, random_state=42)
scores_pr = cross_val_score(model, X_poly, y, cv=kfold, scoring='r2')

# Report on the metrics and output the resultant equation as you did in Part 3.
print("Scores:", scores_pr)
print("Average score: ", np.mean(scores_pr))
print("Standard deviation: ", np.std(scores_pr))

coefficients = model.coef_
intercept = model.intercept_
print(coefficients, intercept)

Scores: [1. 1. 1. 1. 1.]
Average score:  1.0
Standard deviation:  0.0
[ 866.14641337 1032.69506649] -409391.4795834082


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$y=866.15+1032.7x-409391$

Report on the score and interpret: The score among all the folds is perfect, meaning that the model's predictions were correct.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [37]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv('ckd_feature_subset.csv')
ckd_data

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1
...,...,...,...,...,...,...,...
148,0.181818,0.333333,0.075221,0,1,0.457627,0
149,0.207792,0.166667,0.181416,0,1,0.728814,0
150,0.493506,0.166667,0.172566,0,1,0.711864,0
151,0.675325,0.500000,0.128319,0,1,0.745763,0


In [40]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

X_ckd = ckd_data.drop('Target_ckd', axis=1)
y_ckd = ckd_data['Target_ckd']
X_train_ckd, X_test_ckd, y_train_ckd, y_test_ckd = train_test_split(X_ckd, y_ckd, test_size=0.1, random_state=42)

# Logistic Regression
lr_model = LogisticRegression()
lr_model.fit(X_train_ckd, y_train_ckd)
print('Logistic Regression Score:', lr_model.score(X_test_ckd, y_test_ckd))

# Support Vector Machines
svm_model = SVC()
svm_model.fit(X_train_ckd, y_train_ckd)
print('SVM Score:', svm_model.score(X_test_ckd, y_test_ckd))

# k-Nearest Neighbors
knn_model = KNeighborsClassifier()
knn_model.fit(X_train_ckd, y_train_ckd)
print('k-Nearest Neighbors Score:', knn_model.score(X_test_ckd, y_test_ckd))

# Neural Networks
nn_model = MLPClassifier(learning_rate_init=0.1)
nn_model.fit(X_train_ckd, y_train_ckd)
print('Neural Networks Score:', nn_model.score(X_test_ckd, y_test_ckd))

Logistic Regression Score: 0.6875
SVM Score: 0.875
k-Nearest Neighbors Score: 0.9375
Neural Networks Score: 0.9375


## Results and Conclusion for Classification Experiments

In [64]:
# Experiments with Neural Network.
nn_model_1 = MLPClassifier(solver='sgd', learning_rate_init=0.1)
nn_model_1.fit(X_train_ckd, y_train_ckd)
print('Neural Networks Score 1:', nn_model_1.score(X_test_ckd, y_test_ckd))

nn_model_2 = MLPClassifier(solver='sgd', activation='logistic',learning_rate_init=0.01)
nn_model_2.fit(X_train_ckd, y_train_ckd)
print('Neural Networks Score 2:', nn_model_2.score(X_test_ckd, y_test_ckd))

nn_model_3 = MLPClassifier(activation='tanh',learning_rate_init=0.01) # default solver='adam'
nn_model_3.fit(X_train_ckd, y_train_ckd)
print('Neural Networks Score 3:', nn_model_3.score(X_test_ckd, y_test_ckd))

nn_model_4 = MLPClassifier(activation='identity',learning_rate_init=0.1) 
nn_model_4.fit(X_train_ckd, y_train_ckd)
print('Neural Networks Score 4:', nn_model_4.score(X_test_ckd, y_test_ckd))

Neural Networks Score 1: 0.875
Neural Networks Score 2: 0.625
Neural Networks Score 3: 0.875
Neural Networks Score 4: 0.875


## Results and Conclusion for Neural Network Experiments

The best result from the experiments was using the default 'adam' solver (a stochastic gradient-based optimizer), default 'relu' activation function, and an initial learning rate of 0.1 to get an accuracy of 0.9375. From the other experiments, the worst result was from using an sgd solver (stochastic gradient descent), logistic activation function, and initial learning rate of 0.01. Other different configurations of activation functions and use of sgd solver resulted in the same accuracy of 0.875.